graftabledag/graftabledag.go

```diff
@@ -0,0 +1,171 @@
+package graftabledag
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/ipfs/go-cid"
+
+	"github.com/quorumcontrol/chaintree/chaintree"
+	"github.com/quorumcontrol/chaintree/dag"
+)
+
+type GraftableDag interface {
+	GlobalResolve(ctx context.Context, path chaintree.Path) (value interface{}, remaining chaintree.Path, err error)
+	OriginDag() *dag.Dag
+	DagGetter() DagGetter
+}
+
+type DagGetter interface {
+	GetTip(ctx context.Context, did string) (*cid.Cid, error)
+	GetLatest(ctx context.Context, did string) (*chaintree.ChainTree, error)
+}
+
+type GraftedDag struct {
+	dagCache  *lru.Cache // Is this premature optimization?
+	dagGetter DagGetter
+	origin    *dag.Dag
+}
+
+var _ GraftableDag = (*GraftedDag)(nil)
+
+func New(origin *dag.Dag, dagGetter DagGetter) (*GraftedDag, error) {
+	cache, err := lru.New(16)
+	if err != nil {
+		return nil, fmt.Errorf("could not create cache for GraftableDag: %w", err)
+	}
+
+	return &GraftedDag{
+		dagCache:  cache,
+		dagGetter: dagGetter,
+		origin:    origin,
+	}, nil
+}
+
+func (gd *GraftedDag) getChaintreeDag(ctx context.Context, did string) (*dag.Dag, error) {
+	tip, err := gd.dagGetter.GetTip(ctx, did)
```

Comment:
So somewhat related to this, I've had the desire a few times to add an LRU to the tupelo client itself to provide this functionality for both `GetTip` & `GetLatest` (optional obviously, and trivial with func args now) --- the only reason I bring that up is it could limit the `DagGetter` interface here to just `GetLatest`, delegating caching to the client.

Reply:
Yeah, it could delegate to the underlying interface implementer (which won't always be a client). Something to think about...